Symbol aware legend expressions

roya0045 opened the pull request:
# Description
Greetings, this push is in the same vein as most of my previous ones.
The gist of the PR is to allow the user to be able to set up expressions in the legend, yes such a thing is already doable since a recent PR from Mr.Dawson, but this PR seeks to allow greater usability.

Every legend node is evaluated independently. The expression has access to the layer scope, the legend scope, the map scope and the symbol scope.

When using aggregation operations, the symbol scope will be used automatically to filter out the features to use and provide the statistics for the features that are part of the current symbol.

The GUI should allow the user to set the expression of the layertreelayer. The current labels would be accessible as an expression and so the end user could insert an expression anywhere in relation to his label text.

In order to ensure that the counting is done when evaluating the expression, a refresh of the layout legend will be triggered once all counting tasks are completed. In order to do this some signals slots and a list have been implemented in both the vector layer class and the item layout legend class to keep track of all these tasks and force a refresh only when everything is counted.

I have also added more verbose sip error output for debuging ( my sipify script generated bad characters, and I think this verbosity is necessary to match the verbosity of other errors).

### Checklist
[Done] add the expression in the layertreemodellegendnode.

[Done]Set up a trigger somehow so that when the scope item generating the feature IDs per symbol this gets passed to the featurerequest (like a list of selected features get passed to it for evaluating an expression)

[Done]And Change QgsLegendMode::Data to insert the expression?

[Done for python] Unit tests

[Done] Add UI elements

[Done] Find a way to get the expression to be properly evaluated without stalling QGIS.

# Edit
Signals slots and a list were added to keep track of background process ( mostly counting for now) for layers and in the item legend to force a refresh when all tasks are completed. These items could also be used in other cases if desired, possibly with another dedicated container for safety but the current container should be sufficient for most cases,

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

roya0045 commented:
@nyalldawson I'm debating adding the expression in the QgsVectorLayer or in the qgslayermodellegendnode/QgsLegendModel::data (I'm not sure)

On the one hand if the expression is in the VectorLayer it should apply to all it's symbolnode
On the other hand in the legendrelated items it could allow for more customization per element but I would need a way to push all the expression to every embedded node if such a thing is needed.

roya0045 commented:
QUESTION: Would it be a good idea to implement something in the expression parser that would check for FeatureIds and set that in the request when performing aggregate operations? I assume such a thing is done when using a filter in the expression. I am assuming a similar thing is done in the FieldCalculator when using a filter like WITHIN and with selected features.

The goal would be to replicate this behaviour and use the symbolIds to add a filtering element to other filters in the expression (such as filter by map extent).

Anyone has a clue or tip about this?

roya0045 commented:
@nyalldawson please add a squash tag, since I update from the web app it forces me to make a commit for every file changed and this is where there is always too many commits to my pull.

roya0045 commented:
( This is a duplicate of the first Edit):
I've tinkered with the aggregate calculator to provide a list of featureIds as a way of filtering the query.

I have also added a scope check in QgsExpressionFunction, for that step there is something I am unsure about, should I use the subscope mechanic to fetch the symbol scope of the node and get it's IDs and use the legend scope as the main scope or got with the symbol scope as the main scope and ignore the subsetting aspect?

roya0045 commented:
@nyalldawson, @m-kuhn  I get the following error "error: cannot initialize a parameter of type 'QgsExpressionContext *' with an lvalue of type 'const QgsExpressionContext *'"
Usually a sub context is passed there but in the alternate version that I made I pass the pointer to the aggregate method of the VectorLayer. 

Now I'm not sure what would be the best way to handle this, should I const_cast, duplicate the context, or something else?

roya0045 commented:
In regards to the previous commit:
I could have either went with that or attempted to cast the QVariant list outputed by the context to a QgsFeatureIds. I could have also forced the vector layer to accept the symbol id and fetch himself the QgsFeatureIds to provide to the aggregate function but I thought that it would reduce usability and that by leaving the option to input a QgsFeatureIds it was more interesting for future extensions.

I am also not sure if I should force the fids to be const and otherwise I am not sure how to cast the qvariant outputed by the context back to a QgsFeatureIds.

nyalldawson commented:
@roya0045 this is still on my radar, and I'll review as soon as possible. It's likely going to be toward the end of freeze though, as I'm busy bug fixing and focusing on 3.6 for now.

roya0045 commented:
> @roya0045 this is still on my radar, and I'll review as soon as possible. It's likely going to be toward the end of freeze though, as I'm busy bug fixing and focusing on 3.6 for now.

All good thanks! I'll try to iron out the kinks in the meantime.

roya0045 commented:
@m-kuhn @nyalldawson I think I'll do the UI part in another pull since I'm not sure what is the best way to go on about this. I'm thinking about adding another button next to the featurecount button or modifying with a drop-down arrow like with the typical variable buttons. Or overriding it but this might confused some people.

Also about unit tests, I'm not too sure about how to test this properly due to the complexity of it (testing the output in the layoutlegend and with categories/symbol based expressions.

roya0045 commented:
Always evaluating the legend gives me segfault with WMS and makes the legend behave strangely (see https://dash.orfeo-toolbox.org/testDetails.php?test=99765734&build=351036) 

But the layertreetest provided the proper output https://dash.orfeo-toolbox.org/testDetails.php?test=99765681&build=351036, for an expression.

@nyalldawson , @m-kuhn  should I revert the last commit or do I keep it that way and try to find what's breaking?

roya0045 commented:
@nyalldawson , @m-kuhn  looking back at the crash, is it possible that the symbol scope elements would crash elements like wms,wcs,etc. ?
I'll see if there is a way to do a typecheck but if you have a clue about doing this it might be faster to fix.

roya0045 commented:
@elpaso Would you have an idea about what might be the cause of the SEGFAULTS when WMS are used in the tests? I have tried to prevent executing my functions on them and I'm not sure what might be failing ever since the build that passed.

roya0045 commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -530,34 +548,65 @@ void QgsSymbolLegendNode::updateLabel()
-  else if ( role == RuleKeyRole && !mKey.isEmpty() )
-    return mKey;
+  QgsExpressionContext context;
+  if ( mLayerNode->layer()->type() == VectorLayer )
```
I'm not sure about this, can anyone help ( @nyalldawson , @m-kuhn ), I just want to ensure that the source layer is a vector, using vl would mean that the source layer is already casted as a layer and that is not viable.

Do I need to add QgsMapLayer.h in the inclue , if so what would be the syntax to check the returned type, I assume that I should use QgsMapLayer::LayerType::VectorLayer ?

I need to do this since my previous wms check was useless since the wms sources are spatialite for the test and this is what lead to the segfault.

roya0045 commented:
The internal evaluation happens before external evaluation, this means that evaluation externally is useless ( see the last layertreetest) and that some layerscope variables cannot be accessed during evaluation and produce a blank string (see the legend test).

A way to do things would be to evaluate only the layertree/qgsvectorlayer's expression and to leave the general evaluation to the code made by @nyalldawson since there is otherwise two evaluation of the legend expressions and with distinct scope.

I tried to add more scopes/context to the expressioncontext generated by the symbol to allow the use of more scope variables but I may be missing something by only creating the layerscope but I would assume this is sufficient since the layerscope is generated by the utility scope maker.

Currently the option for evaluating the legend expressions could be left to mr Dawson's work and segregate symbol scope evaluation to a custom expression that would apply to all elements of a layer.

But this would also mean that the expression could not be customized for each symbol node.

This behavior needs to be evaluated.

Otherwise I am unsure what is causing the strange legend displays, I feel this aspect needs to be investigated regardless of the previously mentioned elements.

roya0045 committed:
Update qgslayertreemodellegendnode.h

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
adding new functions

roya0045 committed:
adding functions

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
refactoring

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
small changes

roya0045 committed:
updating include

roya0045 committed:
Update qgsvectorlayer.cpp

roya0045 committed:
Update qgsvectorlayer.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
fix attempt

roya0045 committed:
layout

roya0045 committed:
indentation

roya0045 committed:
layout

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
fixing param name

roya0045 committed:
adding new funct

roya0045 committed:
adding new = operator

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
reverting

roya0045 committed:
Update qgsexpressioncontext.h

roya0045 committed:
Update qgsexpressioncontext.cpp

roya0045 committed:
Update qgsexpressioncontext.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
reverting again

roya0045 committed:
reverting

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
typos

roya0045 committed:
forgotten typo

roya0045 committed:
updating sip

roya0045 committed:
updating itemlegend sip

roya0045 committed:
updating modellegendnode sip

roya0045 committed:
reverting test changes

roya0045 committed:
usability improvement

roya0045 committed:
adapting header and adding docs

roya0045 committed:
Update qgslayertreemodellegendnode.sip.in

roya0045 committed:
more reverting

roya0045 commented:
The good news is that the evaluator seem to be able to access the layout scope ( or not at all and it's the legend evaluating it), but somewhere the set custom label doesn't change for some items, strangely.

roya0045 committed:
temporary modification

roya0045 committed:
indentation

roya0045 committed:
indentation

roya0045 committed:
indent

roya0045 committed:
whitespace

roya0045 committed:
docs

roya0045 committed:
passing down the name

roya0045 committed:
debugging

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
more debug tests

roya0045 committed:
trying out a new way

roya0045 committed:
include and layout

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgsexpressioncontext.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
more debug lines

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
typo

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
alternate version

roya0045 committed:
Adding more namefinding code

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
test

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
updating sips1

roya0045 committed:
update sips 2

roya0045 committed:
updating sips 3

roya0045 committed:
Update qgsexpressioncontext.sip.in

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgsexpressioncontext.sip.in

roya0045 committed:
Update qgslayertreemodellegendnode.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
whitespace

roya0045 committed:
changing behavior with existing tmln

roya0045 committed:
changing behavior

roya0045 committed:
layout

roya0045 committed:
typo

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 commented:
@nyalldawson Issues are being solved but I still don't get how the modification made [in this test](https://github.com/roya0045/QGIS/blob/4e4d06b252613b4cbcb016011a415f2e337192ad/tests/src/python/test_qgslayoutlegend.py#L354) don't seem to show the symbol scope item evaluated in the legend. 

The extra space is present but the variable is not displayed, I'm not sure if a condition is causing this or if the scope is not properly initialized...

roya0045 committed:
debug funct

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 committed:
changing debuging

roya0045 commented:
In the previously mentionned test the label is set on the legend node, but when the code is ran, index2legendnode does not fetch the legend node and falls back to using layerLegendNodes.

Since there is no way to know which one is the current legend node when dealing with a layer with multiple symbols, a workaround for index2legendnode should be found or a way to know which layerLegendNode should be used for display.

In that case it should also be considered to not pass the name and fall back on the custom label stored in the legend node. But if index2legendnode stil fails, one might evaluate the wrong node when not all symbols of a layer is displayed and relying on layerLegendNodes...

roya0045 committed:
more debugging

roya0045 committed:
fix build

roya0045 committed:
typo

roya0045 commented:
After the new debugging changes, I don't see [this debug code](https://github.com/roya0045/QGIS/blob/42b67c0dd65dc3e465accf1ae9746783f1117a43/src/core/layout/qgslayoutitemlegend.cpp#L950) pop up, this would mean that despite the fact that the layer has 3 symbol stored,The index2legendNode is not a viable option to fetch the symbols in a legend.

I am confused as to how looping over the symbol node is done since some categories for a given layer may not be present in the legend and would warrant that index to be skipped, or evaluate through some other mean.

But it would seem that the label is applied to the first symbol item and that the other two symbols are left as is. Maybe I forgot something in my pull but shouldn't they be evaluated somehow, unless they are and the proper value is not assigned properly??

roya0045 committed:
Changing the behavior to cover legend nodes

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Merge branch 'master' into proto-legend-exp-1

roya0045 commented:
After the newest method everything seems to be ok.

The main function might be a little heavy as of now but it works. A shorter version could be provided if necessary.

roya0045 committed:
removing debugging

roya0045 committed:
typo

roya0045 committed:
reverting changes

roya0045 committed:
preparing the expression

roya0045 committed:
typo

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
adding unit tests

roya0045 committed:
removing useless prepare expression

roya0045 committed:
fixing test

roya0045 committed:
style

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 commented:
Tests are giving strange results, maybe this is an issue with the counter and I might need to find a way to force the counter to finish before evaluating. 

Also I'm not sure if the trigger in the aggregator is functional but that is dependent on the counter.

@nyalldawson  what would be the best way to postpone the evaluation until the counter is finished?

roya0045 committed:
exposing a variable

roya0045 committed:
force counting

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgsvectorlayer.h

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
Update qgsvectorlayer.cpp

roya0045 committed:
Update qgsvectorlayer.h

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
reverting and typo

roya0045 committed:
removing debug and style

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
attemp to pass sip test

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 commented:
Well not it works, but any sip file that I make doesn't pass the test.

roya0045 committed:
more verbose sip output

roya0045 committed:
fixing bad sipify output

roya0045 committed:
Preparing replacetext expression

roya0045 commented:
@nyalldawson @m-kuhn @nirvn After a lot of trials and tribulations, it works.

I added some extra goodies along the way such as always preparing the expression in the QgsExpression::ReplaceText , the verbose output to the sip test. And possibly some others( can't recall).

There may be a cleanup that needs to be done but I am happy with the results. 

Also please add the squash tag!

I still need to think how I'll tackle the UI part, mostly how to integrate the expression builder to compliment labels or how to take advantage of the expression variable that I added to QgsLayerTreeNode. I don't think the current legenditem buttons can access the QgsLayerTreeNode but I may be mistaken.

I assume we'll have enough time to work out that part after the feature freeze.

roya0045 committed:
Update qgsexpression.cpp

roya0045 committed:
Migrating the expression variable

roya0045 committed:
Update qgslayertreenode.h

roya0045 committed:
Adding expression function

roya0045 committed:
removing expression functions

roya0045 committed:
Update qgslayertreenode.sip.in

roya0045 committed:
Update qgslayertreelayer.sip.in

roya0045 committed:
Update qgslayertreelayer.sip.in

roya0045 committed:
Update qgslayertreenode.sip.in

roya0045 commented:
The UI is done, it is currently in another branch, the UI PR will be created once this one is merged in order to separate those two components ( as I cannot rely on Travis to test the UI part)

roya0045 committed:
test tweaks

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
Expression gui (#7)

* adding new button bindings

* Update qgslayoutlegendwidget.cpp

* add new Icon

* fixing image

* typo

* improved icon

* updating UI

roya0045 commented:
I'm adding the GUI/UI components after-all.

roya0045 committed:
style

roya0045 committed:
typo

roya0045 commented:
Updated description

roya0045 committed:
adding layertreelayer test

roya0045 committed:
indent

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 committed:
Merge branch 'master' into proto-legend-exp-1

roya0045 commented:
@nirvn @m-kuhn  A review would be appreciated from either or both of you, @nyalldawson should review it next week.

roya0045 committed:
manipulation error when resolving conflicts

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
removing foreach

roya0045 committed:
modernizing loop

roya0045 committed:
typo

m-kuhn commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -845,47 +855,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+  if ( replace )
+  {
+    mExpContext.~QgsExpressionContext();
+    QList<QgsExpressionContextScope *> scopes;
```
```suggestion
    const QList<QgsExpressionContextScope *> scopes;
```

Make sure range based for loops are always called on const containers (or they will detach which is an expensive operation)

roya0045 replied:
should I also const [this line?](https://github.com/roya0045/QGIS/blob/a8d3de8bb7b5ef54e9b5a1975ac35351e7d109a2/src/core/qgsvectorlayerfeaturecounter.cpp#L70)

I assume that only returned const objets could be const and that normal things needed to be casted as const. I guess the travis compilation will tell if those items cannot be const.

roya0045 committed:
fixing loop inits

roya0045 committed:
consting lists

m-kuhn commented on src/core/qgsvectorlayerfeaturecounter.cpp:
```diff
@@ -66,9 +68,10 @@ bool QgsVectorLayerFeatureCounter::run()
     {
       renderContext.expressionContext().setFeature( f );
       QSet<QString> featureKeyList = mRenderer->legendKeysForFeature( f, renderContext );
```
```suggestion
      const QSet<QString> featureKeyList = mRenderer->legendKeysForFeature( f, renderContext );
```

m-kuhn commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -539,7 +546,101 @@ void QgsSymbolLegendNode::updateLabel()
+  QVariantList featureIds;
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+
+  QgsFeatureIds fids = counter->featureIds( mItem.ruleKey() );
```
```suggestion
  const QgsFeatureIds fids = counter->featureIds( mItem.ruleKey() );
```

m-kuhn committed:
const loop

Co-Authored-By: roya0045 <roya0045@users.noreply.github.com>

m-kuhn committed:
const loop

Co-Authored-By: roya0045 <roya0045@users.noreply.github.com>

roya0045 commented:
@m-kuhn Thanks for the help! Feel free to review if you have the time.

m-kuhn commented on src/core/qgsvectorlayerfeaturecounter.h:
```diff
@@ -40,22 +42,44 @@ class CORE_EXPORT QgsVectorLayerFeatureCounter : public QgsTask
+
+    /**
+     * Returns the feature Ids for a particular \a legendKey.
+     * If run() has not yet been executed, it will be.
```
it will be [word missing here]?

roya0045 replied:
Bad phrasing, changed it to be clearer.

m-kuhn commented on src/core/qgsvectorlayerfeaturecounter.h:
```diff
@@ -40,22 +42,44 @@ class CORE_EXPORT QgsVectorLayerFeatureCounter : public QgsTask
+     * If run() has not yet been executed, it will be.
+     * If the key has not been found an empty QSet will be returned.
+     *
+     * \since:3.8
```
```suggestion
     * \since QGIS 3.8
```

roya0045 replied:
Done

m-kuhn commented on src/core/qgsvectorlayerfeaturecounter.h:
```diff
@@ -40,22 +42,44 @@ class CORE_EXPORT QgsVectorLayerFeatureCounter : public QgsTask
+     * signal has been emitted.
+     *
+     * \note Not available in Python bindings.
+     * \since: 3.8
```
```suggestion
     * \since QGIS 3.8
```

roya0045 replied:
Done

m-kuhn commented on src/core/qgsvectorlayerfeaturecounter.h:
```diff
@@ -40,22 +42,44 @@ class CORE_EXPORT QgsVectorLayerFeatureCounter : public QgsTask
 
+    /**
+     * Returns the QgsFeatureIds for each symbol. Only valid after the symbolsCounted()
+     * signal has been emitted.
```
Maybe a `\see` linking the two methods would be good since they do (almost) the same?

roya0045 replied:
done

m-kuhn commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -539,7 +546,101 @@ void QgsSymbolLegendNode::updateLabel()
+
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), textOnSymbolLabel(), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  QgsVectorLayerFeatureCounter *counter = vl->countSymbolFeatures( true );
```
Isn't this reducing performance a lot by blocking the UI (even for every single symbol)?
Would it be possible to just use the current counter results (and add a substitution while not available) and update the label when the counter job has finished in the background?

roya0045 replied:
Yes, but I am not sure how to implement such a thing to keep the counting going on and postpone the evaluation.

If you have a suggestion please let me know.

roya0045 replied:
What I can think about is pushing the whole function in QgsApplication::taskManager()->addTask( mFeatureCounter ); and still using the wait,

Or finding another way to delay the evaluation and resume the code execution only when[ the signal from the counter ](https://github.com/roya0045/QGIS/blob/595c9392536d6a17e316280d29f6542b7e4508f0/src/core/qgsvectorlayer.cpp#L725) is emitted. And to maintaining all this is the background to not stall QGIS.

The first option should be somewhat straighforward (I doubt I need to add signal but please correct me) the second one I'm not sure if it's doable. @nyalldawson  any opinion on this?

roya0045 replied:
Signals were used, I'll mark this as resolved

m-kuhn commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -539,7 +546,101 @@ void QgsSymbolLegendNode::updateLabel()
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  // TODO: ADD NECESSARY SCOPES ??
+  context.appendScope( vl->createExpressionContextScope() );
```
https://qgis.org/api/classQgsExpressionContextUtils.html#ae204a266b5a9ae5475f568c264fc719c

roya0045 replied:
I left the TODO for me in case I needed more scopes but I think that by passing down the layout legend+layout map and the context from the vector layer should be sufficient.

I'll remove the todo.

m-kuhn commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -845,47 +855,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
-  return context;
+  if ( replace )
+  {
+    mExpContext.~QgsExpressionContext();
```
What does this do exactly?

roya0045 replied:
This aims to reset the current expression context using the proper method instead of creating a new one and replacing the member.

roya0045 replied:
Moved. to new class

m-kuhn commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -821,7 +831,7 @@ void QgsLayoutItemLegend::onAtlasEnded()
 }
 
-QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+QgsExpressionContext QgsLayoutItemLegend::createExpressionContext( bool replace ) const
```
This method should not modify the internals.
If it's absolutely necessary, a new method `updateExpressionContext()` which is not const should be added.

roya0045 replied:
Since mExpContext is a new internal, and that the expression context is also a new method I thought I should bundle them together.

I have changed things and separated them.

m-kuhn commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -845,47 +855,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+    }
+    else
+    {
+      // QList<QgsLayerTreeModelLegendNode *> legendnodes = nodeLayer->layer()->legend()->createLayerTreeModelLegendNodes( nodeLayer );
```
Remove any leftover code

m-kuhn commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -845,47 +855,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
-    if ( QgsLayerTree::isLayer( node ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) && !node->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
+  QString name;
+  QgsLayerTreeNode *node = index2node( index );
+  QgsLayerTreeModelLegendNode *ltmln = index2legendNode( index ); // Possibly useless
```
Possibly useless?

roya0045 replied:
From the test results and the debugging, it would seem that the index that are iterated over don't provide an index that can be converted into a legend node even for the layers that have a categorical symbology (the jet layer in the legend layout test notably).

m-kuhn commented:
Hi @roya0045 

I had a rough look at it. Some architectural questions can be found in the review so far, apart from some code style comments.

There are still a lot of leftovers in the code (Comments like "TODO" or "Possibly useless" and commented code). Can you review those and either remove or - if you need assistance - add comments with background / context and a question, so we can specifically look at these?

Thanks a lot

roya0045 committed:
Update qgsvectorlayerfeaturecounter.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayertreemodellegendnode.h

roya0045 committed:
Update qgsvectorlayerfeaturecounter.h

roya0045 committed:
creating new method

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -680,6 +683,11 @@ void QgsLayoutItemLegend::setLinkedMap( QgsLayoutItemMap *map )
+
+  // unsure if needed
+  QgsExpressionContext context = updateExpressionContext();
+  mLegendModel->setLayoutExpContext( &mExpContext );
```
I am adding these as a safety measure to ensure that the member mExpContext is kept up to date, is this a good thing to update for every new linked map or should I not do that.

roya0045 commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -844,48 +852,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+    else
+    {
+      const QList<QgsLayerTreeModelLegendNode *> legendnodes = layerLegendNodes( nodeLayer, false );
+      if ( legendnodes.count() > 1 ) // evaluate all existing legend nodes but leave the name for the legend evaluator
```
This replaces the evaluation of every single legend node and parse through every legend node of a given layer since index2legendNode cannot seem to catch and return single legend nodes such as in data-defined legend and other multi-category symbology.

Is this a good idea?

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 commented:
@m-kuhn I have made some changes. I know that I will get sip errors as I cannot solve them currently, I will use the test output to correct them.

I will change the TODO when I find a way to resolve the major issue in qgslayertreemodellegendnode.cpp. Finding a solution to ensure the proper expression evaluation/context creation with up to date information on the feature ids is critical and stalling QGIS was the best shortcut that I could find.

I assume that using a signal would be optimal but I would need some guidance to ensure everything goes well and get triggered at the proper place since anyone printing maps with the expression might run the risk of having some expression not evaluated if the counting process is not completed before the  printing process begin. I assume @nyalldawson  would also have a suggestion on that subject.

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
indent

roya0045 committed:
Update qgsvectorlayerfeaturecounter.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 committed:
Update qgslayertreemodellegendnode.sip.in

roya0045 committed:
Update qgsvectorlayerfeaturecounter.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 commented on src/core/layout/qgslayoutitemlegend.cpp:
```diff
@@ -844,48 +852,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+      name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+    if ( name.isEmpty() )
+      name = node->name();
+
```
@m-kuhn Another option might be to cast the node to a vectorlayer, check for [% or  an expression member (which is done somewhere deeper where the evaluation is done). With this the counting might be initiated here but the issue would be to postpone further steps until the counting is done without stalling QGIS. 

I would assume I would need to make another function to transfer the rest of the code and use the counter with wait to keep the counting and the evaluation in the same thread.

Is this a good idea?

roya0045 committed:
small icon tweaks

roya0045 committed:
Add Tooltip

roya0045 commented:
@m-kuhn I have added the Tooltip for the UI part, should the tag Need-doc be added to the PR?

roya0045 committed:
fixing weird string changes

nyalldawson commented:
Hi @roya0045 

Apologies for the extended delay in reviewing this PR! I've finally had a chance to build and test.

I'd like to focus on the end user logic here before digging into the code changes. I'd suggest that a more flexible approach instead of ALWAYS appending the evaluated expression to the end of the symbol node text we should instead completely replace the symbol node text with the expression. Basically, the expression would act just like a data-defined override for the node text. This would allow users more flexibility in configuring the text, if e.g. they didn't want to put the expression part at the end (but rather the start), OR if they wanted to manipulate the node entry text in some way (title case, truncate, round decimals, ... ?). We could make a new expression default to "@symbol_label" so that it's easy for users to "modify" the current behavior of showing just the label.

What do you think?

roya0045 commented:
> Hi @roya0045
> 
> Apologies for the extended delay in reviewing this PR! I've finally had a chance to build and test.
> 
> I'd like to focus on the end user logic here before digging into the code changes. I'd suggest that a more flexible approach instead of ALWAYS appending the evaluated expression to the end of the symbol node text we should instead completely replace the symbol node text with the expression. Basically, the expression would act just like a data-defined override for the node text. This would allow users more flexibility in configuring the text, if e.g. they didn't want to put the expression part at the end (but rather the start), OR if they wanted to manipulate the node entry text in some way (title case, truncate, round decimals, ... ?). We could make a new expression default to "@symbol_label" so that it's easy for users to "modify" the current behavior of showing just the label.
> 
> What do you think?

Such a thing is already doable since the whole label is evaluated when there are brackets. In my head most use cases the use would simply want to append an expression to every legend element. If they desire more customization they can always do so by changing the text and adding the expression but without the expression builder to help them (I know this is not optimal).

I would agree with you but there are two points that I'm not sure about.

1. The reliance on the symbol_label, this should only return a string if the end user supplied a custom symbol. We would have to create a new function that returns the existing label (check the custom label and them iterate through [the 4 possible storage point of a label for a layer ](https://github.com/qgis/QGIS/blob/e96dce558daab14b8d782270676994c9d8041ee2/src/core/layout/qgslayoutitemlegend.cpp#L907) as it would seem that a text label can be set in many place and that the current behavior is not optimal. Or maybe I'm overlooking something since I am assuming that the custom label can only be set in the legend and that there is no default value set. This would be a great addition in general, I'll see what I can come up with.

2. This would mean that each label would need to be set manually. The current behavior allows the end user to custom label by themselves and as a group by setting a default expression to the parent. I am assuming that you would like something similar but couldn't this lead to confusion if the user sets up and expression and gets confused as to how changed the symbol label once the expression are displayed everywhere.

roya0045 committed:
new funct and optim

roya0045 committed:
new funct declaration

roya0045 commented:
@nyalldawson I have added the new function to get the proper label and used it throughout the code to lighten up the codebase.

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
Update qgslayertreemodellegendnode.cpp

roya0045 committed:
Update qgslayertreemodellegendnode.h

roya0045 commented:
@nyalldawson I just realized something with your proposition, are you proposing to have everything converted to an expression label by default in order to trigger the feature count?

This might cause some slowdown during the legend creation. 

Or simply to have the default text displayed as the label variable in the expression and always add the [% from this point on to the whole legend label?

roya0045 committed:
Update qgslayertreemodellegendnode.sip.in

roya0045 committed:
typo

roya0045 committed:
whitepsace

roya0045 committed:
set default expression

roya0045 committed:
change label behaviour

roya0045 committed:
ensuring eval

roya0045 committed:
tweaking test for new behavior

roya0045 commented:
@nyalldawson I have implemented your suggestion, let me know what you think.

I might have overlooked something but this should work as you suggested in theory.

roya0045 committed:
adding safety net

roya0045 committed:
remove expression brackets from label

roya0045 committed:
indent

roya0045 committed:
indent

roya0045 committed:
clean up only the variable and not the label

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 commented:
@nyalldawson Changing the behavior means that nested expression cannot be evaluated if someone sets the label to have an expression, the expression returned by symbol_label will be expression text that cannot be evaluated, unless I find a way to evaluate twice the code to catch expressions that are stored in the label, which I think would not be optimal.

roya0045 committed:
small tweak

roya0045 commented:
@nyalldawson @m-kuhn This might seem a silly idea but, could I change whatever is calling the data function of the legendmodel to set it in a thread and then update the update the item with a signal. 

Currently the function is used to bind the label to the proper node so I would need a function that refresh the node information once done.

This seems to be overkill to simply ensure that the counting is done before evaluation but if such a thing provide added value it might be nice.

What do you think?

roya0045 commented:
@nyalldawson since you have added expression evaluation at

https://github.com/roya0045/QGIS/blame/24552223e424cfbc91526ccbe9e8122d3fbc187b/src/core/qgslegendrenderer.cpp#L542
and
https://github.com/roya0045/QGIS/blame/24552223e424cfbc91526ccbe9e8122d3fbc187b/src/core/qgslegendrenderer.cpp#L601

And since this should evaluate these elements, do you mind if I were to remove this code since it evaluate code that was already evaluated?

I have implemented this change in my latest commit, I wouldn't mind keeping them there but I doubt such a thing is necessary, please let me know.

@m-kuhn as far as the evaluation goes, I'm thinking about deactivating the wait on the counter and somehow use the signal to force a legend refresh, would that be a good idea?

roya0045 committed:
removing duplicate string evaluation

roya0045 committed:
copy paste error

roya0045 committed:
clearing lists

roya0045 committed:
partial reverting

roya0045 committed:
signals & slots

roya0045 committed:
signals & slots

roya0045 committed:
signals & slots

roya0045 commented:
@m-kuhn I have come up with something using a list to keep track of the counting being done in a layer and for all layers in the legend to trigger a refresh.

What do you think about this solution?

roya0045 committed:
force sleep instead of wait.

roya0045 committed:
typo

roya0045 committed:
typos

roya0045 committed:
indent

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 committed:
adding doneTask

roya0045 committed:
whitespace

roya0045 committed:
whitespace

roya0045 committed:
whitespace

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 committed:
typo

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
Update qgslayoutitemlegend.cpp

roya0045 commented:
Current solution causes segfault, possibly due to the counter not being initiated during evaluation.

roya0045 committed:
adding safety nets

roya0045 committed:
utility function

roya0045 committed:
refactoring

roya0045 committed:
initiatin loop

roya0045 committed:
const loop and safety

roya0045 committed:
initiating list

roya0045 committed:
trying to be safer

roya0045 committed:
Update qgsvectorlayer.h

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
indent

roya0045 committed:
whitespace

roya0045 committed:
Update qgsvectorlayer.h

roya0045 committed:
fixing condition

roya0045 committed:
fixing condition

roya0045 committed:
Update qgsvectorlayer.sip.in

roya0045 committed:
trying to prevent segfault

roya0045 committed:
restoring drawlayertitleinternal

roya0045 committed:
Update test_qgslayoutlegend.py

roya0045 commented:
@nyalldawson in retrospective I'll reimplement your code, what I have done doesn't apply to the layer title when there are subcategories so this acts as a good safeguard when there is no symbol for a given layer text.

roya0045 committed:
test fix

roya0045 commented:
Unrelated test failure.
@m-kuhn when you have the time could you take a look at the code now, the solution seem to work.
@nyalldawson Please tell me if this is what you had in mind as far as the behavior goes.

roya0045 committed:
fixing condition

roya0045 committed:
safety check to prevent segfault

roya0045 committed:
Update qgslayertreelayer.sip.in

roya0045 committed:
Update qgslayertreelayer.h

roya0045 committed:
Update qgslayoutitemlegend.h

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 committed:
Update qgslayoutitemlegend.sip.in

roya0045 commented:
@nirvn could you give a quick review to this PR if you have the time?
@m-kuhn @nyalldawson  should I document more this feature ( and if so where should I do so) or is the current documentation and implementation sufficient ?

roya0045 committed:
adapting typecheck

roya0045 commented:
@nyalldawson I have changed the code to fit the PR where datatype enum are changed and I now have 4 random segfault.

I was thinking about creating a new PR but when I merge master and this branch and try to make a PR from that branch, github indicate that there are thosands of lines of changed code, most of it already merged in master since this PR is only 831 lines of change.

Any idea how I should fix this?

roya0045 commented:
@3nids Would it be possible to remake your enum branch so I can merge it with this one, somehow the tests are using the new enum class and I don't want to merge master in this branch.

m-kuhn commented:
@roya0045 what's your way of working, only on the webinterface or do you have a local checkout?

roya0045 commented:
> @roya0045 what's your way of working, only on the webinterface or do you have a local checkout?

I mostly use the web interface but I have the master and my fork in github desktop on my personnal computer.

I am assuming that merging the branch as is will not create problem in itself as only the necessary code will be changed, but if we insist on passing all the code and if further changes are needed then I asume I will have to correct the tests.

m-kuhn commented:
I thought I could rebase / squash this work in here for you, but if you have a local copy that's going to be troublesome. No worries with merging master then.

In any case, I would *strongly* recommend you get your local development environment working. It will make life easier for you and reviewers.

roya0045 commented:
> I thought I could rebase / squash this work in here for you, but if you have a local copy that's going to be troublesome. No worries with merging master then.
> 
> In any case, I would _strongly_ recommend you get your local development environment working. It will make life easier for you and reviewers.

The local copy is only for forcing merge(this was for my latest PR). You can always rebase/ squash it, I have no issues with that.

About the local dev environment I tried some times ago but encountered too many issues with CMAKE and it not finding dependencies properly. I gave up after some days of fiddling.

m-kuhn commented:
I guess you are using Windows, did you follow this guide?
I used it recently and it worked fine:
https://htmlpreview.github.io/?https://github.com/qgis/QGIS/blob/master/doc/INSTALL.html#toc14

roya0045 commented:
@m-kuhn I tried to use VSCode instead of VS, this might be the issue or I might have overlooked something.

Other than that if the new code is fine with integration with master, are you ok with the changes that I made to ensure an update of the legend (once counting is done), instead of stalling QGIS ?

nyalldawson commented:
@roya0045 can I strongly encourage you to keep trying to get a local build working? A big part of the delay in reviewing this pr is due to the lack of logical commits with descriptive log messages. Using the web editor is ok for one or two fixes, but this current approach makes reviewing super hard.

I'd suggest if you can't get the build working in windows, to try setting up a Ubuntu virtual machine in VirtualBox. Then getting a build setup is a matter of copying a few commands direct from the install guide.

roya0045 commented:
@nyalldawson Yes if you go through each commit instead of checking the global changes this will be horrible. I'll try to rebase the master of my fork, and to open a new PR with this feature or try to change the branch in the coming hours.

nyalldawson commented:
Thanks at @roya0045 ! 

I suggest the next logical step to do is to see if there's any single portion of the PR we can break off into an atomic step toward your end goal. Can we "chunk off" any parts of this for individual quick review and merge?

roya0045 commented:
> Thanks at @roya0045 !
> 
> I suggest the next logical step to do is to see if there's any single portion of the PR we can break off into an atomic step toward your end goal. Can we "chunk off" any parts of this for individual quick review and merge?

I would like to keep it all together. What I suggest is that I'll keep trying to get my repo fixed ( I've spent the whole night trying to get things to work but my things are so out of date or some other random things are causing a lot of conflicts and I'm trying to avoid fixing thousands of files)

If I can get this in a single block, possibly in a new PR I'll add more documentation ( not in the code directly) to help reviewing the code).

roya0045 commented:
@nyalldawson I tried to merge the current master to a new branch from my master, and I still get some weird artefact when it comes times to compare that branch and the current master. Its possible that rebasing will solve my problem but rebasing is problematic even if my branch was just merged with master.

nyalldawson commented:
@roya0045 

Here's what I'd do in this situation --

1. I'd make a new branch of master
2. "merge and squash" your existing branch to the new branch "git merge --squash"
3. do a "soft reset" to "uncommit" the changes: `git reset --soft HEAD^`, followed up `git reset` to unstage everything.
4. Use a set of `git add -i` / git commit steps to interactively add the individual parts of your work as individual logical commits, with descriptive messages for each commit describing the rationale behind each change

Does that make sense to you? It's a bit of command line git work, but while learning git can be frustrating, it's definitely NEVER a wasted effort :stuck_out_tongue_winking_eye:

m-kuhn commented:
> @nyalldawson Yes if you go through each commit instead of checking the global changes this will be horrible. I'll try to rebase the master of my fork, and to open a new PR with this feature or try to change the branch in the coming hours.

That is a very good move, thanks!
But still: there are likely more commits to come and the chances to get this merged in a reasonable time will be *much* higher if you can get a local build env to work. I'm sure if you spend time on this now, you will compensate for it easily.

elpaso commented:
@roya0045 if you are looking for a ready-to-go dev environment for QGIS you might want to have a look to https://github.com/elpaso/qgis-dev-vagrant, I prepared it for the HF workshop and it should have all you need to build and debug QGIS using a VM.

NathanW2 commented:
Oh sweet. That is pretty cool and would be super good for helping people
get started

On Thu., 28 Mar. 2019, 9:49 pm Alessandro Pasotti, <notifications@github.com>
wrote:

> @roya0045 <https://github.com/roya0045> if you are looking for a
> ready-to-go dev environment for QGIS you might want to have a look to
> https://github.com/elpaso/qgis-dev-vagrant, I prepared it for the HF
> workshop and it should have all you need to build and debug QGIS using a VM.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/8911#issuecomment-477524215>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3JaozQp0qZmoUWFPuexCr0lOp1W0ks5vbJI7gaJpZM4aJbp7>
> .
>

roya0045 commented:
@nyalldawson This is what I meant to do with rebasing my master but the issue is that my current master contains update of masters that were squashed, so I 'became' the author of that PR, even when reversing that PR, it only creates a new PR. So when it comes time to rebase rebasing goes through every PR even if that PR was reverted. This would mean that I would have to fix the same file many times to pass each step of the rebasing process. (Using git reset --hard is of no help)

The optimal solution would be for me to push this branch in the repo of someone else while I purge my fork and reintegrate this branch. Squashing should occur somewhere during those transfers.

I only need someone that would be willing to hold my branch for a little time while I clean up my stuff and merge back this branch.

Even with a dev environment I cannot fathom a way to solve the issues of too many commit while not breaking anything for the next PR.
